### PR TITLE
fix i18n script

### LINKED ISSUE #17: Enhance i18n
**Description Problem**
PR [#6](https://github.com/rawadhossain/talawa-admin/pull/6) successfully implements the core i18n detection infrastructure (`scripts/check-i18n.js`), which identifies hardcoded user-visible text in the codebase. However, during testing on the production codebase, we've identified false positives that will block legitimate development work when the pre-commit hook is active.

### Common False Positive Patterns (few examples)

The current script incorrectly flags (there can be other incorrect flags as well) the following as violations:

1. **TypeScript type annotations**
   ```tsx
   const handler = (): string => { ... }  // Flags "string" as violation
   const fn = (): Promise<void> => { ... }  // Flags "void" as violation
   ```

2. **CSS utility classes in `className` attributes**
   ```tsx
   <div className={`btn primary m-3 p-2`} />  // Flags "btn", "primary"
   <i className="fi fi-rr-home" />  // Flags font icon classes
   ```

3. **Technical ARIA attributes**
   ```tsx
   <div aria-hidden="true" role="button" data-testid="my-test" />
   ```

4. **Date format strings**
   ```tsx
   <input placeholder="YYYY-MM-DD" />  // Technical format, not user text
   ```

5. **Regular expression patterns**
   ```tsx
   const pattern = "[a-z]+";  // Technical pattern, not user text
   ```

6. **Developer-facing messages**
   ```tsx
   console.log("Debug info");  // Developer message, not user-facing
   throw new Error("Internal error");  // Error message, not UI text
   ```

7. **GraphQL queries, JSON operations, string methods**
   ```tsx
   const query = gql`query { ... }`;
   JSON.stringify({ key: "value" });
   ```

8. **JavaScript comparison operators in JSX**
   ```tsx
   {age >= 18 && age <= 40}  // Logical comparison, not text
   ```

## Related
- Prerequisite: PR #6 (i18n detection infrastructure)
- i18n detection Issue: Issue https://github.com/rawadhossain/talawa-admin/issues/5

### PR DESCRIPTION
**Pull request by rawadhossain**
**What kind of change does this PR introduce?**

**Issue Number:**
Fixes #17 

Related PR https://github.com/rawadhossain/talawa-admin/pull/6
**Snapshots/Videos:**

**If relevant, did you update the documentation?**

**Summary**

**Does this PR introduce a breaking change?**

### CodeRabbit AI Review
- [ ] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [ ] I have implemented or provided justification for each non-critical suggestion
- [ ] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [ ] I have written tests for all new changes/features
- [ ] I have verified that test coverage meets or exceeds 95%
- [ ] I have run the test suite locally and all tests pass

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

## Summary by CodeRabbit

* **New Features**
  * i18n checker more accurately avoids false positives by skipping non-user-visible attributes, URLs, dates, regexes, code-like contexts, complex className patterns, and honoring ignore-comment hints.
  * Added a fixture demonstrating common false-positive patterns to validate analyzer behavior.

* **Tests**
  * Expanded and reorganized test coverage for ignore-comments, date/URL/regex detection, className heuristics, and edge cases.
  * Added test utilities for creating temporary test files and running the checker.

* **Chores**
  * Removed an obsolete example fixture.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/check-i18n.test.js [hunk 1]
```diff
@@ -506,4 +582,504 @@ describe('check-i18n script', () => {
...
+        tmp,
+        'api-endpoint.tsx',
+        // FIX: Split to multiple lines so the script can separate the URL from the Text
+        '<a href="api/v1/users">\n  View Users\n</a>',
+      );
+      const res = runScript([file]);
+      expect(res.status).toBe(0);
+      expect(res.stdout).toContain('No non-internationalized');
+      expect(res.stdout).not.toContain('api/v1/users');
+    });
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/check-i18n.test.js]: @coderabbitai this test allows API endpoint patterns in href but flags multi-word JSX text fails when the coderabbit suggestion is applied:
expect(res.status).toBe(1);
expect(res.stdout).toContain('View Users');

recommend the changes to do to pass this test, check the script src/check-i18n.test.js if needed

[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/check-i18n.test.js]: @coderabbitai it still failed. however i made these changes and it pass, is my change correct:
```
it('allows API endpoint patterns in href but flags multi-word JSX text', () => {
      const tmp = makeTempDir();
      const file = writeTempFile(
        tmp,
        'api-endpoint.tsx',
        '<a href="api/v1/users">\n  View Users\n</a>',
      );
      const res = runScript([file]);
      expect(res.status).toBe(1);
      expect(res.stdout).toContain('View Users');
      expect(res.stdout).not.toContain('api/v1/users');
    });
```

script `check-i18n.js` change:
```
// Multi-line JSX text detection
    // Check if current line is text content between JSX tags (not on same line as tags)
    if (idx > 0 && idx < lines.length - 1) {
      const prevLine = lines[idx - 1];
      const nextLine = lines[idx + 1];
      const trimmedLine = line.trim();

      // Check if we're between JSX tags:
      // - Previous line ends with > (opening tag)
      // - Current line has text (not empty, not starting with <)
      // - Next line starts with </ (closing tag) or current line ends with </
      const prevEndsWithTag = />\s*$/.test(prevLine);
      const nextStartsWithClosingTag = /^\s*<\//.test(nextLine);
      const lineEndsWithClosingTag = /<\//.test(line);
      const isTextLine =
        trimmedLine.length > 0 &&
        !trimmedLine.startsWith('<') &&
        !trimmedLine.startsWith('{') &&
        !trimmedLine.startsWith('//') &&
        !trimmedLine.startsWith('/*');

      if (
        prevEndsWithTag &&
        isTextLine &&
        (nextStartsWithClosingTag || lineEndsWithClosingTag)
      ) {
        // This line is JSX text content between tags
        const text = trimmedLine;
        const matchIndex = line.indexOf(text);

        // Skip if in a context that should be ignored
        if (!isInSkipContext(line, matchIndex)) {
          // Skip if this looks like JavaScript code
          if (
            !/(>=|<=|==|!=|===|!==|&&|\|\|)\s*\d+/.test(text) &&
            !/\d+\s*(>=|<=|==|!=|===|!==|&&|\|\|)/.test(text) &&
            !/(return|const|let|var|if|while|for)\s+.*(>=|<=|==|!=|===|!==)/.test(
              prevLine,
            )
          ) {
            if (!isAllowedString(text)) {
              violations.push({ line: lineNumber, text });
            }
          }
        }
      }
    }
```

File: src/check-i18n.test.js [hunk 2]
```diff
@@ -506,4 +582,504 @@ describe('check-i18n script', () => {
...
+        tmp,
+        'comparison-ops.tsx',
+        '<div>{age >= 18 && age <= 40 ? `Adult Person` : `Minor Person`}</div>',
+      );
+      const res = runScript([file]);
+      expect(res.status).toBe(0);
+      expect(res.stdout).toContain('No non-internationalized');
+      // But should not flag the comparison operators themselves
+      expect(res.stdout).not.toContain('>= 18 && age');
+    });
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/check-i18n.test.js]: @coderabbitai test fails after chaining as per your suggestion. however i made these changes and it pass, is my change correct:
```
 it('skips comparison operators in JSX but flags user text', () => {
      const tmp = makeTempDir();
      const file = writeTempFile(
        tmp,
        'comparison-ops.tsx',
        '<div>{age >= 18 && age <= 40 ? `Adult Person` : `Minor Person`}</div>',
      );
      const res = runScript([file]);
      expect(res.status).toBe(1);
      expect(res.stdout).toContain('Adult Person');
      expect(res.stdout).toContain('Minor Person');
      expect(res.stdout).not.toContain('>= 18 && age');
    });
```
script `check-i18n.js` change:
```
 // Only process template literals that are inside JSX expressions (after {)
      // or are clearly JSX content (not in strings, comments, etc.)
      const beforeMatch = line.substring(0, matchIndex);
      // Check if this template literal is inside a JSX expression { ... }
      // Look backwards for an opening brace { that hasn't been closed
      let braceCount = 0;
      let foundJsxExpression = false;
      for (let i = beforeMatch.length - 1; i >= 0; i--) {
        if (beforeMatch[i] === '}') braceCount++;
        else if (beforeMatch[i] === '{') {
          if (braceCount === 0) {
            foundJsxExpression = true;
            break;
          }
          braceCount--;
        }
      }

      // Skip if not in a JSX expression context (unless it's clearly user-visible)
      if (!foundJsxExpression) {
        // Still check if it's in a JSX attribute context
        const afterMatch = line.substring(matchIndex + templateMatch[0].length);
        if (!/^\s*=/.test(afterMatch) && !/=\s*$/.test(beforeMatch)) {
          continue;
        }
      }
```

File: src/check-i18n.test.js [hunk 3]
```diff
@@ -46,38 +46,110 @@ afterEach(() => {
...
+    const fileClean = writeTempFile(
+      tmp,
+      'edge-cases-clean.tsx',
+      `
+       <div><img src="https://example.com" /><span>123</span></div>
+    `,
     );
+    const res = runScript([fileClean]);
+    expect(res.status).toBe(0);
   });
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/check-i18n.test.js]: Addressed and new approach implemented in commit 4c7783e8f853af3f3edc2a6d2e902a2cf370a5e9

File: src/check-i18n.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/check-i18n.test.js]: Addressed and new approach implemented in commit 4c7783e8f853af3f3edc2a6d2e902a2cf370a5e9

File: src/check-i18n.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/check-i18n.test.js]: Addressed and new approach implemented in commit 4c7783e8f853af3f3edc2a6d2e902a2cf370a5e9

File: src/check-i18n.test.js [hunk 4]
```diff
@@ -506,4 +582,504 @@ describe('check-i18n script', () => {
...
+    });
+  });
+
+  describe('Comprehensive false positives fixture', () => {
+    it('passes for false-positives.tsx with all edge cases', () => {
+      const res = runScript([path.join(fixturesDir, 'false-positives.tsx')]);
+      expect(res.status).toBe(0);
+      expect(res.stdout).toContain('No non-internationalized');
+    });
+  });
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/check-i18n.test.js]: Addressed and new approach implemented in commit 4c7783e8f853af3f3edc2a6d2e902a2cf370a5e9
This issue has already been addressed. The false-positives.tsx fixture was updated earlier, and lines 56–57 no longer contain user-visible strings like "Adult" or "Minor". They now include only JavaScript comparison expressions, which are intentionally skipped by the checker.

I’ve verified that the script reports “No non-internationalized user-visible text found” and the test passes for false-positives.tsx with all edge cases passes as expected.

This comment appears to be based on an earlier version of the fixture, so no further changes are needed.

File: scripts/__fixtures__/false-positives.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+      {/* Technical attributes - should be skipped */}
+      <div data-testid="my-test" role="button" aria-hidden="true" />
+
+      {/* URL patterns - should be skipped */}
+      <a href="orgstore/id=123" />
+      <a href={`orgstore/id=${dialogRedirectOrgId}`} />
+      <a href="api/v1/users" />
+      <img src="/assets/logo.png" alt="" />
+      <link href="data:image/png;base64,abc" />
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply scripts/__fixtures__/false-positives.tsx:53]: This file is an intentionally minimal test fixture used to validate i18n false-positive detection, not a production component or accessibility example. The markup here is deliberately incomplete so we can exercise skip logic for technical attributes and URL patterns without introducing additional user-visible or translatable text.

Adding tabIndex or aria-label would change the semantics of the fixture and introduce text that the checker is designed to flag, so I’m keeping this as-is.

File: scripts/check-i18n.js [hunk 6]
```diff
@@ -97,32 +314,220 @@ const collectViolations = (filePath) => {
...
+      }
+
+      // Skip if not in a JSX expression context (unless it's clearly user-visible)
+      if (!foundJsxExpression) {
+        // Still check if it's in a JSX attribute context
+        const afterMatch = line.substring(matchIndex + templateMatch[0].length);
+        if (!/^\s*=/.test(afterMatch) && !/=\s*$/.test(beforeMatch)) {
+          continue;
+        }
+      }
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply scripts/check-i18n.js:496]: Agreed that the JSX expression detection logic is a bit complex. The current approach is intentional to avoid false positives without introducing a full AST parse, and it’s behaving correctly based on the current test coverage.

File: src/check-i18n.test-utils.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  fs.writeFileSync(filePath, content);
+  return filePath;
+};
+
+export const cleanupTempDirs = () => {
+  tempDirs.forEach((dir) => {
+    fs.rmSync(dir, { recursive: true, force: true });
+  });
+  tempDirs.length = 0;
+};
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/check-i18n.test-utils.js:45]: Since this is a test utility, cleanup is intentionally best-effort and silent. fs.rmSync with { recursive: true, force: true } already handles missing directories without throwing, and adding logging here would introduce noise without improving test reliability.

I’ll keep the current implementation as-is.

**Comment by rawadhossain** [comment]
@coderabbitai full review and reevaluate all resolved items and include the following points:
1. Validate whether any updated test files:    
    1. Cover all use cases
    2. Have valid tests
    3. Provide 100% test code coverage
    4. Require any additional tests
 2. Identify:
    1. Lines of code that are missing tests.
    2. Additional tests that may be required
    3. Any unnecessary files that have been submitted and ask that they be excluded.
 3. Validate that the PR:
    1. Meets the requirements of the linked issue.
    2. Complies with all previously suggested improvements and comments.
 
If the submitted code requires any changes, even trivial ones, then make coderabbitai[bot] give the review a CHANGES_REQUESTED status. If no changes are required then give the PR an APPROVED status.

File: src/check-i18n-enhanced.test.js [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+    });
+  });
+
+  describe('Comprehensive false positives fixture', () => {
+    it('passes for false-positives.tsx with all edge cases', () => {
+      const res = runScript([path.join(fixturesDir, 'false-positives.tsx')]);
+      expect(res.status).toBe(0);
+      expect(res.stdout).toContain('No non-internationalized');
+    });
+  });
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/check-i18n-enhanced.test.js]: getAttributeName is currently exercised indirectly through the higher-level attribute detection tests, and the new comprehensive fixture is intended to cover real-world combinations rather than function-level behavior. Given that this is a script-style tool and not a library API, I’ve intentionally kept tests focused on observable behavior to avoid over-coupling tests to internal helpers.

That said, I agree focused unit tests could be useful if this helper grows more complex in the future. I’ll keep this in mind, but for now I’d prefer to keep the test suite aligned with end-to-end behavior.

File: src/check-i18n-enhanced.test.js [hunk 9]
```diff
@@ -0,0 +1,506 @@
...
+      const file = writeTempFile(
+        tmp,
+        'promise-type.tsx',
+        'const handler = async (): Promise<void> => { };',
+      );
+      const res = runScript([file]);
+      expect(res.status).toBe(0);
+      expect(res.stdout).toContain('No non-internationalized');
+    });
+  });
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/check-i18n-enhanced.test.js]: Check the last commit and tell has it been resoled?
The mixed.tsx fixture is used in check-i18n-basic.test.js, so it's not unused

**Comment by rawadhossain** [comment]
@coderabbitai full review and reevaluate all resolved items and include the following points:
1. Validate whether any updated test files:    
    1. Cover all use cases
    2. Have valid tests
    3. Provide 100% test code coverage
    4. Require any additional tests
 2. Identify:
    1. Lines of code that are missing tests.
    2. Additional tests that may be required
    3. Any unnecessary files that have been submitted and ask that they be excluded.
 3. Validate that the PR:
    1. Meets the requirements of the linked issue.
    2. Complies with all previously suggested improvements and comments.
 
If the submitted code requires any changes, even trivial ones, then make coderabbitai[bot] give the review a CHANGES_REQUESTED status. If no changes are required then give the PR an APPROVED status.

**Comment by rawadhossain** [comment]
@coderabbitai full review and reevaluate all resolved items and include the following points:
1. Validate whether any updated test files:    
    1. Cover all use cases
    2. Have valid tests
    3. Provide 100% test code coverage
    4. Require any additional tests
 2. Identify:
    1. Lines of code that are missing tests.
    2. Additional tests that may be required
    3. Any unnecessary files that have been submitted and ask that they be excluded.
 3. Validate that the PR:
    1. Meets the requirements of the linked issue.
    2. Complies with all previously suggested improvements and comments.
 
If the submitted code requires any changes, even trivial ones, then make coderabbitai[bot] give the review a CHANGES_REQUESTED status. If no changes are required then give the PR an APPROVED status.

**Comment by rawadhossain** [comment]
@coderabbitai Has this PR covered comprehensive enhancement of `check-i18n` script along with the tests?
Is it now future proof of any other potential false positives across the codebase? I want it to be complete, robust and devoid of any bugs or potential future issues. Assure me is it ready to merge?

**Comment by rawadhossain** [comment]
@coderabbitai full review and reevaluate all resolved items and include the following points:
1. Validate whether any updated test files:    
    1. Cover all use cases
    2. Have valid tests
    3. Provide 100% test code coverage
    4. Require any additional tests
 2. Identify:
    1. Lines of code that are missing tests.
    2. Additional tests that may be required
    3. Any unnecessary files that have been submitted and ask that they be excluded.
 3. Validate that the PR:
    1. Meets the requirements of the linked issue.
    2. Complies with all previously suggested improvements and comments.
    3. Complete, robust and devoid of any bugs or potential future issues.
 
If the submitted code requires any changes, even trivial ones, then make coderabbitai[bot] give the review a CHANGES_REQUESTED status. If no changes are required then give the PR an APPROVED status.

**Comment by rawadhossain** [comment]
@coderabbitai
Does this suggestion really needs to be addressed as it is shown or my current file setup is perfectly okay as we have done the full review a couple of times?

**Overly broad regex pattern detection may cause false negatives.**
 
 `looksLikeRegexPattern` returns `true` if the text contains ANY of the common regex metacharacters (`^$.*+?{}()[]|\`). This will match legitimate user-visible text containing parentheses, periods, or question marks — e.g., `"Hello (world)"`, `"What?"`, `"Dr. Smith"`.
 
 The second condition on line 204 is especially permissive and could suppress valid violations.
 
 Consider tightening the heuristic to require multiple regex-specific characters or specific patterns:
 
```diff
  const looksLikeRegexPattern = (text) => {
    const trimmed = text.trim();
 -  // Common regex patterns (character classes, quantifiers, anchors, etc.)
 -  if (/^[.*+?^${}()|[\]\\\/\-]+$/.test(trimmed)) return true;
 -  // Patterns with common regex syntax
 -  if (/[\^$.*+?{}()[\]|\\]/.test(trimmed)) return true;
 +  // Only pure regex syntax (no alphabetic characters)
 +  if (/^[.*+?^${}()|[\]\\\/\-]+$/.test(trimmed)) return true;
+  // Character classes like [a-z], [A-Z0-9], etc.
 +  if (/\[[^\]]*[a-z]-[a-z][^\]]*\]/i.test(trimmed)) return true;
 +  // Regex quantifiers attached to patterns like \d{4}, \w+
 +  if (/\\[dDwWsS][*+?{]/.test(trimmed)) return true;
    return false;
  };
 ```

**Comment by rawadhossain** [comment]
@coderabbitai there will be a problem with the tests files if i move them from `src\` to some folder like `src\test-utils` right? I tried and all the same tests failed.

i am asking you think because maintainer asked - "move the tests to a folder under src/test-utils where they are better placed"

**Comment by rawadhossain** [comment]
@coderabbitai current this is the file below, what changes do i make here for moving to subdirectory from `src/check-i18n.test-utils.js` to `src/test-utils/check-i18n.test-utils.js`:
```
import path from 'path';
import fs from 'fs';
import os from 'os';
import { spawnSync } from 'child_process';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export const scriptPath = path.resolve(
  __dirname,
  '..',
  'scripts',
  'check-i18n.js',
);
export const fixturesDir = path.resolve(
  __dirname,
  '..',
  'scripts',
  '__fixtures__',
);

const tempDirs = [];

export const runScript = (targets, options = {}) => {
  const { env, ...rest } = options;
  const res = spawnSync(process.execPath, [scriptPath, ...targets], {
    encoding: 'utf-8',
    env: { ...process.env, ...(env ?? {}), FORCE_COLOR: '0', NO_COLOR: '1' },
    timeout: 30_000,
    ...rest,
  });
  if (res.error) throw res.error;
  return res;
};

export const makeTempDir = () => {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'i18n-detector-'));
  tempDirs.push(dir);
  return dir;
};

export const writeTempFile = (dir, relPath, content) => {
  const filePath = path.join(dir, relPath);
  fs.mkdirSync(path.dirname(filePath), { recursive: true });
  fs.writeFileSync(filePath, content);
  return filePath;
};

export const cleanupTempDirs = () => {
  tempDirs.forEach((dir) => {
    fs.rmSync(dir, { recursive: true, force: true });
  });
  tempDirs.length = 0;
};
```

**Comment by rawadhossain** [comment]
@coderabbitai
Does this suggestion really needs to be addressed as it is shown or my current file setup is perfectly okay as we have done the full review a couple of times?

 **Incorrect order of template literal interpolation removal causes false positives.**
  The regex on line 553 (`/\$\{[^}]*\}/g`) removes simple interpolations but breaks when encountering nested template literals. The pattern `[^}]*` stops at the FIRST `}`, which splits nested templates like `${`inner ${x} more`}`:
 
 1. Line 553 incorrectly matches `${`inner ${x}`(stops at`}`in`${x}`)
 2. Result: ` more`}` (broken structure)
 3. Line 555's nested-template pattern can no longer match
 4. Final static text: ` more`}` instead of empty string
 
 This causes false positives when template literals contain only nested interpolations.
 
 Apply this diff to remove nested template literals BEFORE simple interpolations:
 
 ```diff
        // Strip out variables FIRST (including nested template literals)
        // Remove ${...} patterns, but be careful with nested backticks
        let staticText = fullText;
 -      // Remove simple ${var} patterns
 -      staticText = staticText.replace(/\$\{[^}]*\}/g, '');
        // Remove nested template literals ${`...`}
        staticText = staticText.replace(/\$\{[^`]*`[^`]*`[^}]*\}/g, '');
 +      // Remove simple ${var} patterns
 +      staticText = staticText.replace(/\$\{[^}]*\}/g, '');
        staticText = staticText.trim();
 ```

**Comment by rawadhossain** [comment]
@coderabbitai in the script I am asked to make this required changes below: 
guide me what changes to make in the `scripts/check-i18n.js` add the missing JSDoc Documentation?
### ❌ Required Changes
**1. Missing JSDoc Documentation** The main script `scripts/check-i18n.js` lacks JSDoc comments for its helper functions. Per the Documentation & Code Comments requirements (section 3.2), adequate JSDoc comments are needed on public functions and complex logic.

**Functions requiring JSDoc:**

* `walk(dir)` - Directory traversal function
* `shouldAnalyzeFile(filePath)` - File filtering logic
* `stripComments(content)` - Comment removal
* `hasIgnoreComment(originalLines, lineIndex)` - Ignore directive detection
* `countWords(text)` - Unicode-aware word counting
* `looksLikeUrl(text)` - URL pattern detection
* `looksLikeDateFormat(text)` - Date format detection
* `looksLikeRegexPattern(text)` - Regex pattern detection
* `isInSkipContext(line, matchIndex)` - Context-aware skip logic
* `isAllowedString(text)` - String validation
* `toPosixPath(filePath)` - Path normalization
* `getAttributeName(line, matchIndex)` - Attribute extraction
* `collectViolations(filePath)` - Core violation detection
* `main()` - Entry point

Each function should have a JSDoc comment describing:

* Purpose/description
* @param tags with types and descriptions
* @returns tag with type and description
* @example for complex functions where helpful
